Content.Client/Tabletop/TabletopSystem.cs

```diff
@@ -0,0 +1,303 @@
+using Content.Client.MobState;
+using Content.Client.Stunnable;
+using Content.Client.Tabletop.Components;
+using Content.Client.Tabletop.UI;
+using Content.Client.Viewport;
+using Content.Shared.Interaction.Helpers;
+using Content.Shared.Tabletop;
+using Content.Shared.Tabletop.Events;
+using JetBrains.Annotations;
+using Robust.Client.GameObjects;
+using Robust.Client.Input;
+using Robust.Client.Player;
+using Robust.Client.UserInterface;
+using Robust.Client.UserInterface.CustomControls;
+using Robust.Shared.GameObjects;
+using Robust.Shared.GameStates;
+using Robust.Shared.Input;
+using Robust.Shared.Input.Binding;
+using Robust.Shared.IoC;
+using Robust.Shared.Map;
+using Robust.Shared.Maths;
+using DrawDepth = Content.Shared.DrawDepth.DrawDepth;
+
+namespace Content.Client.Tabletop
+{
+    [UsedImplicitly]
+    public class TabletopSystem : SharedTabletopSystem
+    {
+        [Dependency] private readonly IInputManager _inputManager = default!;
+        [Dependency] private readonly IUserInterfaceManager _uiManger = default!;
+        [Dependency] private readonly IPlayerManager _playerManager = default!;
+
+        // Time in seconds to wait until sending the location of a dragged entity to the server again
+        private const float Delay = 1f / 10; // 10 Hz
+
+        private float _timePassed; // Time passed since last update sent to the server.
+        private IEntity? _draggedEntity; // Entity being dragged
+        private ScalingViewport? _viewport; // Viewport currently being used
+        private SS14Window? _window; // Current open tabletop window (only allow one at a time)
+        private IEntity? _table; // The table entity of the currently open game session
+
+        public override void Initialize()
+        {
+            CommandBinds.Builder
+                        .Bind(EngineKeyFunctions.Use, new PointerInputCmdHandler(OnUse, false))
+                        .Register<TabletopSystem>();
+
+            SubscribeNetworkEvent<TabletopPlayEvent>(OnTabletopPlay);
+            SubscribeLocalEvent<TabletopDraggableComponent, ComponentHandleState>(HandleComponentState);
+        }
+
+        public override void Update(float frameTime)
+        {
+            // If there is no player entity, return
+            if (_playerManager.LocalPlayer is not { ControlledEntity: { } playerEntity }) return;
+
+            if (StunnedOrNoHands(playerEntity))
+            {
+                StopDragging();
+            }
+
+            if (!CanSeeTable(playerEntity, _table))
+            {
+                StopDragging();
+                _window?.Close();
+                return;
+            }
+
+            // If no entity is being dragged or no viewport is clicked, return
+            if (_draggedEntity == null || _viewport == null) return;
+
+            // Make sure the dragged entity has a draggable component
+            if (!_draggedEntity.TryGetComponent<TabletopDraggableComponent>(out var draggableComponent)) return;
```

Comment:
You seem to be doing the expensive check first and the cheaper checks after?

Reply:
That is actually correct, as `CanSeeTable` check has to go before `_draggedEntity` check (window should close if not in range even when not dragging anything), `StunnedOrNoHands` check should go before `CanSeeTable` as it is cheaper (does not contain `InRangeUnobstructed`).